src/main/java/org/alfresco/repo/search/impl/parsers/AlfrescoFunctionEvaluationContext.java

```diff
@@ -499,6 +499,87 @@ else if(field.endsWith(QueryConstants.FIELD_TRANSFORMATION_TIME_SUFFIX))
+            PropertyDefinition propertyDef = dictionaryService.getProperty(propertyField);
+            if (!propertyDef.getDataType().getName().equals(DataTypeDefinition.DATETIME))
+            {
+                throw new FTSQueryException(QueryConstants.FIELD_SOLR_UNIT_OF_TIME_HOUR_SUFFIX+" only supported on datetime properties");
+            }
+            else
+            {
+                return propertyField;
+            }
+        }
+        else if(field.endsWith(QueryConstants.FIELD_SOLR_UNIT_OF_TIME_DAY_SUFFIX))
+        {
+            QName propertyField = QName.createQName(field.substring(0, field.length() - QueryConstants.FIELD_SOLR_UNIT_OF_TIME_DAY_SUFFIX.length()));
+            PropertyDefinition propertyDef = dictionaryService.getProperty(propertyField);
+            if (!propertyDef.getDataType().getName().equals(DataTypeDefinition.DATE) &&
+                    !propertyDef.getDataType().getName().equals(DataTypeDefinition.DATETIME))
+            {
+                throw new FTSQueryException(QueryConstants.FIELD_SOLR_UNIT_OF_TIME_DAY_SUFFIX+" only supported on date and datetime properties");
+            }
+            else
+            {
+                return propertyField;
+            }
+        }
+        else if(field.endsWith(QueryConstants.FIELD_SOLR_UNIT_OF_TIME_MONTH_SUFFIX))
+        {
+            QName propertyField = QName.createQName(field.substring(0, field.length() - QueryConstants.FIELD_SOLR_UNIT_OF_TIME_MONTH_SUFFIX.length()));
+            PropertyDefinition propertyDef = dictionaryService.getProperty(propertyField);
+            if (!propertyDef.getDataType().getName().equals(DataTypeDefinition.DATE) &&
+                    !propertyDef.getDataType().getName().equals(DataTypeDefinition.DATETIME))
+            {
+                throw new FTSQueryException(QueryConstants.FIELD_SOLR_UNIT_OF_TIME_MONTH_SUFFIX+" only supported on date and datetime properties");
+            }
+            else
+            {
+                return propertyField;
+            }
+        }
+        else if(field.endsWith(QueryConstants.FIELD_SOLR_UNIT_OF_TIME_YEAR_SUFFIX))
+        {
+            QName propertyField = QName.createQName(field.substring(0, field.length() - QueryConstants.FIELD_SOLR_UNIT_OF_TIME_YEAR_SUFFIX.length()));
+            PropertyDefinition propertyDef = dictionaryService.getProperty(propertyField);
+            if (!propertyDef.getDataType().getName().equals(DataTypeDefinition.DATE) &&
+                    !propertyDef.getDataType().getName().equals(DataTypeDefinition.DATETIME))
+            {
+                throw new FTSQueryException(QueryConstants.FIELD_SOLR_UNIT_OF_TIME_YEAR_SUFFIX+" only supported on date and datetime properties");
+            }
+            else
+            {
+                return propertyField;
+            }
+        }
```

Comment:
Probably an initial if with `field.contains("_unit_of_time_")` or so, encapsulating all the different cases could save some execution time.